### ISSUE TITLE
Relationship fields don't update after related entry is saved

### ISSUE THREAD

**Issue by swierczek** [ExpressionEngine/ExpressionEngine#4059]
**Description of the problem**
I have a relationship field that only accepts Open entries, and I have 4 Open entries selected. If I edit one of those related entries and Close it, the entry doesn't get removed from the relationship field, and `{field_name:total_results}` still outputs 4, even though only 3 of them are open and the field is configured to only allow Open entries.

Additionally, I can even save the original entry, even though I'd expect it to fail validation on the relationship field now that a non-Open entry is selected. I'd expect that entry to either automatically get removed from the field, or validation to fail so the invalid related entry would need to be edited before saving successfully.

![image](https://github.com/ExpressionEngine/ExpressionEngine/assets/2423727/2eee3d4d-9b1b-4cbe-91a7-81f284a475bb)

```
<h1>outside of loop: {testing_relate:total_results}</h1>
{testing_relate status="open"}
	<h2>count: {testing_relate:count}</h2>
	<h2>inside of loop: {testing_relate:total_results}</h2>
	<p>{testing_relate:title}</p>
{/testing_relate}
```

<img width="890" alt="image" src="https://github.com/ExpressionEngine/ExpressionEngine/assets/2423727/d432b874-16ec-4fc6-a657-6a7e39050d81">

Interestingly, if I remove the `outside of loop` line from the template, then the inside `:total_results` is at least correct (3). The first way (`:total_results` outside of the loop) is possibly incorrect due to some caching/shortcut happening in [Parser.php](https://github.com/ExpressionEngine/ExpressionEngine/blob/7.dev/system/ee/legacy/libraries/relationship_parser/Parser.php#L151) when `:total_results` is parsed first outside of the loop:

```
{testing_relate status="open"}
	<h2>count: {testing_relate:count}</h2>
	<h2>inside of loop: {testing_relate:total_results}</h2>
	<p>{testing_relate:title}</p>
{/testing_relate}
```

<img width="866" alt="image" src="https://github.com/ExpressionEngine/ExpressionEngine/assets/2423727/b25faf5f-2a38-4310-8e06-25df7c2cc8a2">

TL;DR - I'd expect `:total_results` to only ever output `3` in this state, because the field is configured to only allow Open entries, and closing 1 of the 4 Open entries should now act as if only 3 entries are in the field.

**Environment Details:**
 - Version: EE 7.3.15

**Comment by intoeetive** [comment]
@swierczek thank you for the detailed report!

There are 2 things here, 

- one is the inconsistent output of `:total_results` (I think 3 should be correct number in your case), and this is a bug

- the other is whether the relationships should be automatically updated when related entry status is changed, and I would say the current behavior is intended. The status setting affects what you see as available for selection and it does not affect the data that's already stored. In fact, if we change the data automatically without user making explicit action, that might be confusing and even considered as data loss in some cases.

**Comment by intoeetive** [comment]
I had a look and it seems that it's not possible to apply `status="open"` parameter to relationship **single** tag with `:total_results` modifiers. Some limitation of relationship parser. So to get correct count, one needs to use tag pair with status parameter and loops... which is not ideal.

**Comment by swierczek** [comment]
Agreed with your assessment of the 2 parts to this. I guess for the second part, I'd at least expect the validation to fail when attempting to save the entry, but I think it makes sense to _not_ remove it from the field/database automatically.

Good idea with the workaround, but I potentially run into the same issue, I think due to parsing caching.

This works as expected (outputs "3 entries"):

```
{if "{testing_relate status='open'}{if testing_relate:count == 1}{testing_relate:total_results}{/if}{/testing_relate}" == "4"}
	4 entries
{if:else}
	3 entries
{/if}
```

But the following outputs "outside of loop: 4" and "4 entries":

```
<h1>outside of loop: {testing_relate:total_results}</h1>
{if "{testing_relate status='open'}{if testing_relate:count == 1}{testing_relate:total_results}{/if}{/testing_relate}" == "4"}
	4 entries
{if:else}
	3 entries
{/if}
```

But if we change the "outside of loop" logic to the same loop, it works ("new outside of loop: 3" and "3 entries"):

```
<h1>new outside of loop: {testing_relate status='open'}{if testing_relate:count == 1}{testing_relate:total_results}{/if}{/testing_relate}</h1>
{if "{testing_relate status='open'}{if testing_relate:count == 1}{testing_relate:total_results}{/if}{/testing_relate}" == "4"}
	4 entries
{if:else}
	3 entries
{/if}
```

I hope this helps!
